### ISSUE TITLE
Create a simple Scylla troubleshooting tool

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#8631]
We have issues like #8586 (forbid SimpleStrategy in CREATE KEYSPACE) where we suggest that certain operations or options often cause problems in real deployment where they are used unintentionally. There are also certain data modeling mistakes - such as ending up with large rows, extremely large partitions, tons of tombstones, and so on - which users might not be aware of but are hurting their workload's performance or stability.

In this issue I want to propose an external troubleshooting tool, probably written in Python, which connects to a cluster via CQL and checks for known potential problems. Using the above examples, the tool can check whether any of the existing keyspaces use SimpleStrategy and can check how large cell/row/partition checking is configured and whether it reported anything. The tool will also do many other sanity checks that Scylla's support people usually do when approaching an unknown malfunctioning deployment. It is debatable whether this tool should have only access to CQL, to metrics, or also to logs (a harder problem on a cluster) - we can start with CQL only and consider expanding to metrics and logs later. The tool can also ask Scylla to log a sample of the queries and investigate whether any of the queries are problematic, too slow, and so on.

The result of the tool's run will be a report listing _potential_ problems with the setup of the database or its content, with explanations on why each problem was reported, what its implications might be, and how to fix this problem (ideally, we would just have web pages on troubleshooting specific problems, that the tool will point to).

In the past, instead of the the external troubleshooting tool I propose here, it was proposed that Scylla itself would have its own internal troubleshooting logic, which will periodically print messages to the log on potential problems. I argue that the approach I propose here, based on an external tool, is better, because:
1. It will be easier to write and quickly improve this Python tool than in-Scylla C++ code.
2. It will be possible for Scylla support as well as individual users to run improved version of the tool on older installations, without upgrading Scylla itself.
3. Using a tool to scan the _current_ situation and report it now, is easier than rummaging through logs looking for warnings.
4. With warnings in the log, there is always a conflict of interest between showing fewer of them when you are satisfied with the cluster's operation, and showing more of them when the cluster is misbehaving and you're trying to understand why.
5. Warnings are about a single event or check. A single configuration problem can lead to many different or identical warning or error messages, and it is hard for inexperienced users to reach conclusions on what they did wrong.

CC @eliransin

**Comment by tzach** [comment]
@nyh this seems similar to exiting node_health_check script, maybe we can enhance it
See also #4378 

The tool can also include a self reporting feature, opening a Scylla issue and update relevant info in it

**Comment by amnonh** [comment]
And has similiar functionality to the Advisor,
You can create Advisor specific issues here:
https://github.com/scylladb/scylla-monitoring/issues/new?assignees=&labels=advisor&template=50-advisor-issue.md

**Comment by nyh** [comment]
@tzach,

I looked at `node_health_check` now (I have to admit I didn't know it existed), and while it is a script, like I propose, the actual philosophy and implementation is **very different**: 

node_health_check` collects a lot of information into a "report" (information dump), with the intention that a human then looks at all this information and tries to figure out potential problem. But this human needs to be an experienced support person, who has seen all the common problems and how they manifest in various configuration options, CQL results, and so on. I don't think this is realistic or scalable as the project grows.

Instead, I propose that support people convert their thinking into code (or write a document and let developers convert the idea into code).

As a trivial example, `node_health_check` today just prints out all the tables schemas. Presumably, a support person would then look at this output and look for fishy things, like usage of the SimpleStrategy replication strategy. So this is exactly what the code can do - check if SimpleStrategy was used, and if it was, report the potential problem (with a text or link explaining the risk of using this strategy). If an experienced support person considers 7 different things in the schema - our code will check all 7.

I'm not suggesting that we can solve every problem this way. Instead I'm hoping to 1. save an hour of the experienced support person's time when he or she approaches a new problem which might be an obvious one. 2. allow new, **inexperienced**, support people to help users without referring simple problems to more experienced support people or developers. 3. perhaps allow users to help themselves - without support people.

**Comment by tzach** [comment]
@nyh yes agree with everything you wrote.

@amnonh 
The advantages of using the existing motioning system Advisor are:
- a centralize element
- rule and alert engine, which is easy to integrate externally 
- have access to logs and CQL

Maybe we can build a simple scylla_exporter which will report the strategy per keyspace, and let the alert manager to report on SimpleStrategy. And maybe scylla_exporter can be implemented in Scylla, or in the monitoring stack (using CQL)

**Comment by nyh** [comment]
I'm not familiar with the Advisor, until today I didn't know it had access to CQL and logs - I thought it was just about the metrics.
If our chosen way is Advisor, I think another issue we should consider is whether we can upgrade it indepedently of Scylla. I can see us wanting to use the same up-to-date advisor with hot-off-the-presses checks invented yesterday by our best support engineers - on an older installation of Scylla.

**Comment by nyh** [comment]
Recently we had a customer issue where some external process was stealing CPU time from Scylla and causing stall and high tail latencies. It took us quite time until one of us (as usual, @avikivity) guessed that what is causing the high tail latencies was stalls, and what was causing stalls was an external process. The troubleshooting tool I am suggesting here could have detected this situation in 10 seconds.

It means this tool should probably be able to use "ssh" in addition to the other tools I mentioned earlier (metrics, logs, stall detection). Or instead of a general ssh (which can be a security hole), some sort of agent belonging to the troubleshooting tool which will be able to extract information (e.g., output of "ps") from the nodes.

**Comment by vladzcloudius** [comment]
> Recently we had a customer issue where some external process was stealing CPU time from Scylla and causing stall and high tail latencies. 

Stalls are **always** a source of high tail latency. This is a no-brainer. The more frequent the stalls - the lower the latency percentile impacted by it. This is a well known fact.

> It took us quite time until one of us (as usual, @avikivity) guessed that what is causing the high tail latencies was stalls, and what was causing stalls was an external process. The troubleshooting tool I am suggesting here could have detected this situation in 10 seconds.

You need to know what to look for, @nyh. Then you can program that tool to do the same. 

Here the problem wasn't the stall. The stall was a side effect.
The problem was the CPU hogging external process. And assuming that only those that consume 100% of the CPU are the issue is wrong. Any process running together with scylla on the same CPU which is active all the time and thereby causing scylla thread to context switches all the time is going to impact latencies.

So, no "magic tool" is going to reliably help.
What would helped to identify the issue faster would be a scylla (seastar) metric showing how much time other than a seastar threads were using the same CPU where a seastart thread is bound to.

Having such a metric would allow anybody to simply throw a glimpse at the corresponding graph and know exactly what the problem is.

**Comment by nyh** [comment]
@vladzcloudius your logic is wrong - you are saying the same thing I did and then conclude it can't be done :-( 

First of all, as you said, an automated script looking at the log could say "it's a stall". You called this a "no-brainer" but some of the people you have for debugging real-world issues (including your's truely) are worse than having no brain - they have no experience. I'm not sure how much time it would have taken me to figure this out myself - both to figure out that I should look for a stall, and then how to look for it. I'm not saying I couldn't have done this - but I don't know how much time it would have taken me. Not all of us when seeing megabytes of logs realize this immediately. As you said in another thread, Avi has diagnosed dozens of these cases so he knows exactly how to look at them. A tool, with special code written to look for stalls, could have also said that "Found stalls!". Not as one of 100 different metrics, not a 100 graphs with a million points of information - but one message, that even a no-brainer or inexperienced debugger like myself would have understood.

Second, Avi used "ps" or similar and noticed that another process was running on the same CPU as Scylla. Yes, we could have added a 101st metric in Seastar that measures something similar, but still a person would have needed to think to look at that metric and realize what this metric means. Avi thought to do this after a long time that several other smart people didn't. A tool would have tirelessly checked for this condition among 100 other conditions programmed in, and in a few seconds exclaimed "found another process, 'dirmgr', running on the same CPU as Scylla!".

By the way, you're right that a metric over time might have been better than running "ps", because it's possible that this dirmgr doesn't run all the time and only wakes up once an hour (say), so a running "ps" at the wrong time might not have found it, but a graph over time would - seeing non-zero spikes. That's exactly the role of the human experts - like you and Avi - to realize we should prepare this graph. Then the role of the robotic expert (the tool I'm proposing) would be to analyze this graph automatically and say "I see here a non-zero spike - that's bad". Not every human non-expert could have looked at 100 different graphs and realize that graph #73 when it has a non-zero spike means trouble.

I'm not proposing a "magic" tool. I'm not proposing GPT-level AI that will replace Avi. I am proposing that you, Avi, and other people who keep seeing the same problems over and over, write code that detects these problems instead of keeping the detection "algorithm" in your own brain and being frustrated when other people don't know these mental algorithms.

> Having such a metric would allow anybody to simply throw a glimpse at the corresponding graph and know exactly what the problem possibly is.

This is so wrong that I don't even know where to start. You don't understand how frustrating it is for me when you tell me to "look at the metrics, there is a problem". The first problem is that "metrics" is plural - there isn't "a graph", there are dozens and dozens of graphs. I don't know which of those to look at, and how to recognize a bad-looking shape from a good-looking shape. If I'm not sure what a specific graph measures, there is no good documentation to look at and of course no written wisdom on how to recognize problems in specific graphs. Vlad, if you want to keep being angry that only Avi and a couple more people always solve real-world issues and keep blaming the rest of the developers for not caring or being lazy or I don't know what, you can do that. But I'm trying to suggest a better way.

One day I'll just stop listening to the nay-sayers on this and similar threads and just sit on my ass and code such a tool. One of the reasons why I haven't started doing it so far (besides my managers always telling me to work on other things) is that I don't think I have enough knowledge to put **into** such a tool myself :-(

> But in general, as mentioned above (https://github.com/scylladb/scylladb/issues/8631#issuecomment-946866633) SD is a tool that is supposed to "check the sanity" of a scylla node.

I don't want a tool that only checks the static "sanity" of the node - i.e., detect that a node is configured with "suspicious" options. Such a tool is indeed important, **before** trouble happens. I want a tool that checks for ongoing or recent problems in the dynamic behavior of the system - including logs, ps, metrics, cql queries, and so on. I want it to check the same things that the experts (you, Avi, etc.) check when told about a problem, and report to the developer specific things he should look at (e.g., a specific graph, a specific process, and so on).

**Comment by vladzcloudius** [comment]
> @vladzcloudius your logic is wrong - you are saying the same thing I did and then conclude it can't be done :-(

Before claiming that somebody's logic is "wrong" you should probably make sure you understood what that person's "logic" was. ;)

And in my case it was: "No matter how good 'the tool' is going to be it still won't be able to replace an experienced engineer that would apply his experience to all collected info - including the one from that tool. It doesn't mean however that we should not try to improve tools like Scylla Doctor or Scylla Monitoring to help more with cases we have already seen in the past."

See more below.

> 
> First of all, as you said, an automated script looking at the log could say "it's a stall". You called this a "no-brainer" but some of the people you have for debugging real-world issues (including your's truely) are worse than having no brain - they have no experience. I'm not sure how much time it would have taken me to figure this out myself - both to figure out that I should look for a stall, and then how to look for it. I'm not saying I couldn't have done this - but I don't know how much time it would have taken me. Not all of us when seeing megabytes of logs realize this immediately. As you said in another thread, Avi has diagnosed dozens of these cases so he knows exactly how to look at them. A tool, with special code written to look for stalls, could have also said that "Found stalls!". Not as one of 100 different metrics, not a 100 graphs with a million points of information - but one message, that even a no-brainer or inexperienced debugger like myself would have understood.

Nadav, I guess you realize that "Found stalls!" message will not be the only message that 'the tool' is going to be printing, right? It would still take a person to understand if that's the one he/she should pay attention to.

But, you know what? You are right! Stalls are important. That's why we have visualized them in Scylla Monitoring long-long time ago (see a screenshot below). 

> 
> Second, Avi used "ps" or similar and noticed that another process was running on the same CPU as Scylla. Yes, we could have added a 101st metric in Seastar that measures something similar, but still a person would have needed to think to look at that metric and realize what this metric means. 

Well, there are ways to make one metric more visible than others in our Monitoring.
However it all starts with adding a metric. ;)

> Avi thought to do this after a long time that several other smart people didn't. A tool would have tirelessly checked for this condition among 100 other conditions programmed in, and in a few seconds exclaimed "found another process, 'dirmgr', running on the same CPU as Scylla!".

It was 'top' but anyway...

> 
> By the way, you're right that a metric over time might have been better than running "ps", because it's possible that this dirmgr doesn't run all the time and only wakes up once an hour (say), so a running "ps" at the wrong time might not have found it, but a graph over time would - seeing non-zero spikes. 

Exactly! ;)

> That's exactly the role of the human experts - like you and Avi - to realize we should prepare this graph. Then the role of the robotic expert (the tool I'm proposing) would be to analyze this graph automatically and say "I see here a non-zero spike - that's bad". Not every human non-expert could have looked at 100 different graphs and realize that graph #73 when it has a non-zero spike means trouble.

You would be surprised to hear that we have heard you and have already implemented this. ;)

For example see what we have on our "Overview" dashboard. Check this out:

![image](https://github.com/scylladb/scylladb/assets/6103981/ed314f0e-7dc3-42f1-945f-0f9a2cbffc4f)

I also encourage you to check our "Scylla CQL" dashboard.

> 
> I'm not proposing a "magic" tool. I'm not proposing GPT-level AI that will replace Avi. I am proposing that you, Avi, and other people who keep seeing the same problems over and over, write code that detects these problems instead of keeping the detection "algorithm" in your own brain and being frustrated when other people don't know these mental algorithms.

We do it, Nadav. For quite some time. But you didn't know because you probably don't use our Monitoring frequently enough. And this one of my main points all along.

> 
> > Having such a metric would allow anybody to simply throw a glimpse at the corresponding graph and know exactly what the problem possibly is.

> 
> This is so wrong that I don't even know where to start. You don't understand how frustrating it is for me when you tell me to "look at the metrics, there is a problem". The first problem is that "metrics" is plural - there isn't "a graph", there are dozens and dozens of graphs. I don't know which of those to look at, and how to recognize a bad-looking shape from a good-looking shape. If I'm not sure what a specific graph measures, there is no good documentation to look at and of course no written wisdom on how to recognize problems in specific graphs. 

There is a Help message at almost every graph. Check this out:

![image](https://github.com/scylladb/scylladb/assets/6103981/0d673dbc-595c-4b87-b916-4163b6be3a40)

Some have a more detailed description than others. You are welcome to check them all out and file a GH issue if you'd like a better description.

But no description is going to help unless you use the Monitoring.
And no magic tool is going to replace a person **who has too know** which shape of a particular graph is good and which isn't in general. For trivial cases we have already implemented quick checkers as I have already mentioned above.

And, whether you like it or not there will be more than a single (or even a few) metrics to look at. Our system is complicated - it can't be described by one or two (or even 10) values.

> Vlad, if you want to keep being angry that only Avi and a couple more people always solve real-world issues and keep blaming the rest of the developers for not caring or being lazy or I don't know what, you can do that. But I'm trying to suggest a better way.

I'm not angry. I hope you are not angry either. Being angry is counterproductive in general. 
I'm just trying to make sure we are both on the same page. 

> 
> One day I'll just stop listening to the nay-sayers on this and similar threads and just sit on my ass and code such a tool. One of the reasons why I haven't started doing it so far (besides my managers always telling me to work on other things) is that I don't think I have enough knowledge to put **into** such a tool myself :-(

One wise person once told me: "Before starting writing something new (SW) make sure it's not already written. There is a 99% probability that it already is." 

In this particular case it's 100% probability ;)

> 
> > But in general, as mentioned above ([#8631 (comment)](https://github.com/scylladb/scylladb/issues/8631#issuecomment-946866633)) SD is a tool that is supposed to "check the sanity" of a scylla node.
> 
> I don't want a tool that only checks the static "sanity" of the node - i.e., detect that a node is configured with "suspicious" options. Such a tool is indeed important, **before** trouble happens. 

Just a note: such a distinction is arguable. SD is valuable not less **after** trouble happened.

However this is of course irrelevant to the current discussion.

> I want a tool that checks for ongoing or recent problems in the dynamic behavior of the system - including logs, ps, metrics, cql queries, and so on. I want it to check the same things that the experts (you, Avi, etc.) check when told about a problem, and report to the developer specific things he should look at (e.g., a specific graph, a specific process, and so on).

We already wrote it. It's called Scylla Monitoring. And it already has some of these implemented. We continuously working on improving it. If you have specific ideas of your own you are welcome to start with at least filing an RFC with a **concrete** suggestion towards the corresponding repository (`scylla-monitoring`).

cc @amnonh

**Comment by nyh** [comment]
@vladzcloudius, I understood you well. You seem to believe that every one of the Scylla developers has enough experience (say, 1/10th of the experience that you and Avi have) in reading monitoring graphs, and all of us immediately see stalls and all sorts of other problems when faced with 50 colorful graphs with mysterious names. This is patently false, as you can see by looking at almost any issue which involves looking monitoring. I wish it was just me that is an idiot who can't solve any issue, but you can issues where multiple developers spent days looking at various graphs without understanding what those graphs mean or what the problem is, until somebody more experienced comes along and says: "oh, it's obvious! It's a stall!" (or one of the 20 other "obvious" things).

You can continue thinking that Scylla developers are lazy idiots and if we just force them to spend more time reading graphs, they will all become experts like you. You can continue thinking that our monitoring is self explanatory and its "help" is useful (while in fact, I am never able to understand anything from it, and see also https://github.com/scylladb/scylladb/issues/8888). But continuing to think this will not solve any of our problems. You can continue thinking (like some of the other managers think, apparently) that if we just force all developers to spend 1 hour or 2 hours or 8 hours every day looking at customer bugs and monitoring, they will become better at it - which may be true but will leave them less time to do real development, and more worryingly - will hurt these developers sense of accomplishment and self-worth (it's not a great feeling spending days on an issue that you can't solve because you don't have the necessary experience).

Again, I don't think that an automatic tool will solve every problem in every case. I do think it will be able to solve some of them, quickly, probably many of them more quickly than I can.

Finally, if you want to convince me that every problem can be solved in the context of Scylla monitoring (like you seemed to suggest now), please, you and Avi tell me that you always look just at the monitoring - not at logs, not at "top", not at CQL queries, not at the configuration - nothing. If the answer is "no, of course not, I need all these things to understand issues", well, it means that the monitoring is _not_ enough and we can't expect all the problems to be visible in it, which is what you tried to claim above.

**Comment by vladzcloudius** [comment]
@nyh When I say that I think you need to try spending more time with Monotoring in order to understand it better doesn't mean that I think you or any other Scylla developer is a "lazy idiot". I re-read my posts a few times to understand what made you think so but it still left me puzzled.

> Finally, if you want to convince me that every problem can be solved in the context of Scylla monitoring (like you seemed to suggest now), please, you and Avi tell me that you always look just at the monitoring - not at logs, not at "top", not at CQL queries, not at the configuration - nothing. If the answer is "no, of course not, I need all these things to understand issues", well, it means that the monitoring is not enough and we can't expect all the problems to be visible in it, which is what you tried to claim above.

90% of the issues are solved using Monitoring and SD.

The rest 10% of "interesting cases" - no tool is going to help you because, like that issue that CPU hogger application, they require "outside the box" thinking.

So, my previous point stands: we should continue improving our existing tools.
It doesn't mean we (or me as you seem to imply) should ignore your inputs.

If you think that our tools (e.g. Monitoring) are not developer-friendly this falls into the "let's improve them" category. Me and Amnon would love to hear your ideas as to possible improvements. Amnon is usually very quick to implement pilots.

For example, if you think that some graph's explanation is unclear you can:
1) Find which scylla metric stands behind it.
2) Figure out what is the actual semantics of it by either reading the code yourself or asking the developer who added it.
3) Suggest a better description in a form of GH issue to `scylla-monitoring`. 

I'm 100% sure in a matter of single digit days your proposal is going to be merged and the next Monitoring release is going to have it.

The same goes to SD: if you have a specific idea as to what check can be automated do share. I'll do my best to get in implemented ASAP.

**Comment by mykaul** [comment]
@nyh I'm not sure what's missing from the combination of Scylla Doctor (and perhaps Scylla Doctor Cluster?) and ScyllaDB Monitoring, for a proper troubleshooting tool? 
One example I could think of is https://github.com/scylladb/field-engineering/tree/master/scylla-scripts/streaming-progress , as well as any other log mining tooling, what else?
I think Scylla Doctor could be extended to fit most use cases. Analyzing metrics is one area which I feel might be challenging for such a tool.

**Comment by nyh** [comment]
@mykaul, according to the above discussion:
1. "Scylla Doctor is only meant to check the host configuration.". It doesn't and won't (IIUC) check logs, metrics, traces, "describe schema", "ps", or anything else except misconfigurations.
2. Scylla Monitoring doesn't help you figure out problems you're not already an expert in. And it also doesn't check logs, query traces, describe, ps, etc.

**Comment by mykaul** [comment]
> @mykaul, according to the above discussion:
> 
> 1. "Scylla Doctor is only meant to check the host configuration.". It doesn't and won't (IIUC) check logs, metrics, traces, "describe schema", "ps", or anything else except misconfigurations.

It's Python script... We can extend it.

> 2. Scylla Monitoring doesn't help you figure out problems you're not already an expert in. And it also doesn't check logs, query traces, describe, ps, etc.

To some extent, I agree (wrt metrics). Some can give you a good hint though. SD can certainly be extended to do more than it does today - checking logs, for example.

**Comment by vladzcloudius** [comment]
> @mykaul, according to the above discussion:
> 
>     1. "Scylla Doctor is only meant to check the host configuration.". It doesn't and won't (IIUC) check logs, 

It collects them - doesn't analyze them yet however.

> metrics,

Metrics are supposed to be analyzed and visualized by Monitoring - not SD.

> traces, 

This is also clearly not a node-local info.
Traces analysis is a totally independent topic. We should start with implementing Traces offloading and visualization like Zipkin.

> "describe schema", 

It does collects and analyzes it.

> "ps", or anything else except misconfigurations.

I think I've already commented enough above why analyzing `top` (not `ps`! :D) makes no sense in SD.



> 
>     2. Scylla Monitoring doesn't help you figure out problems you're not already an expert in. 

It very much can. 

> And it also doesn't check logs, query traces, describe, ps, etc.

See above.

**Comment by nyh** [comment]
Another idea of something to check in the troubleshooting tool - missing NTP:

If the clocks of the different nodes in the cluster are not synchronized, various problems can happen (do we have this documented anywhere?) - one we saw recently in a customer is what @avikivity predicted in his review of https://github.com/scylladb/scylladb/pull/12529 - parallel aggregation breaks when the clocks are not synchronized, because the wrong timeouts are used and some of the nodes will just timeout immediately.

The troubleshooting tool should be able to detect that the clocks of the different nodes aren't synchronized. This can be done in various ways, with increasing preference in my opinion:
1. Using ssh to inquire the current time on each nodes, and comparing them.
2. Investigate the processes running on the nodes, looking for NTP.
3. Using REST API to inquire the current time on each node. We don't have such a REST API yet (we only have /system/uptime).
4. Using CQL to inquire the current time on each node, using `select toTimestamp(now()) from system.local;` (thanks [stackoverflow](https://stackoverflow.com/questions/29272691/get-current-date-in-cassandra-cql-select))

Methods 3-4 can't detect small problems, but on a fast LAN should easily detect inconsistency of time that are more than a few milliseconds.

**Comment by tarzanek** [comment]
I think we will need to use some of the ideas here for reporting tool about cluster health @ivenn-scylladb (e.g. combining metrics to get some conclusion about state of cluster)
